metadata/src/lib.rs

```diff
@@ -2,101 +2,619 @@
+//! 2. Obtaining [`frame_metadata::RuntimeMetadataPrefixed`], and then
+//!    using `.try_into()` to convert it into [`Metadata`].
+
+#![deny(missing_docs)]
+
+mod from_into;
+mod utils;
+
+use scale_info::{form::PortableForm, PortableRegistry, Variant};
+use std::collections::HashMap;
+use std::sync::Arc;
+use utils::ordered_map::OrderedMap;
+use utils::variant_index::VariantIndex;
+
+type ArcStr = Arc<str>;
+
+pub use from_into::TryFromError;
+pub use utils::validation::MetadataHasher;
+
+/// Node metadata. This can be constructed by providing some compatible [`frame_metadata`]
+/// which is then decoded into this. We aim to preserve all of the existing information in
+/// the incoming metadata while optimizing the format a little for Subxt's use cases.
+#[derive(Debug, Clone)]
+pub struct Metadata {
+    /// Type registry containing all types used in the metadata.
+    types: PortableRegistry,
+    /// Metadata of all the pallets.
+    pallets: OrderedMap<ArcStr, PalletMetadataInner>,
+    /// Find the location in the pallet Vec by pallet index.
+    pallets_by_index: HashMap<u8, usize>,
+    /// Metadata of the extrinsic.
+    extrinsic: ExtrinsicMetadata,
+    /// The type ID of the `Runtime` type.
+    runtime_ty: u32,
+    /// The type Id of the `DispatchError` type, which Subxt makes use of.
+    dispatch_error_ty: Option<u32>,
+    /// Details about each of the runtime API traits.
+    apis: OrderedMap<ArcStr, RuntimeApiMetadataInner>,
+}
+
+impl Metadata {
+    /// Access the underlying type registry.
+    pub fn types(&self) -> &PortableRegistry {
+        &self.types
+    }
+
+    /// Mutable access to the underlying type registry.
+    pub fn types_mut(&mut self) -> &mut PortableRegistry {
+        &mut self.types
+    }
+
+    /// The type ID of the `Runtime` type.
+    pub fn runtime_ty(&self) -> u32 {
+        self.runtime_ty
+    }
+
+    /// The type ID of the `DispatchError` type, if it exists.
+    pub fn dispatch_error_ty(&self) -> Option<u32> {
+        self.dispatch_error_ty
+    }
+
+    /// Return details about the extrinsic format.
+    pub fn extrinsic(&self) -> &ExtrinsicMetadata {
+        &self.extrinsic
+    }
+
+    /// An iterator over all of the available pallets.
+    pub fn pallets(&self) -> impl ExactSizeIterator<Item = PalletMetadata<'_>> {
+        self.pallets.values().iter().map(|inner| PalletMetadata {
+            inner,
+            types: self.types(),
+        })
+    }
+
+    /// Access a pallet given its encoded variant index.
+    pub fn pallet_by_index(&self, variant_index: u8) -> Option<PalletMetadata<'_>> {
+        let inner = self
+            .pallets_by_index
+            .get(&variant_index)
+            .and_then(|i| self.pallets.get_by_index(*i))?;
+
+        Some(PalletMetadata {
+            inner,
+            types: self.types(),
+        })
+    }
+
+    /// Access a pallet given its name.
+    pub fn pallet_by_name(&self, pallet_name: &str) -> Option<PalletMetadata<'_>> {
```

Comment:
As my previous comment, it would nice to return Result<PalletMetadata<'_>>, PalletNameNotFound> here to avoid doing it when it's called.

Reply:
I think `Option`'s are cleaner here, and then the thing using the metadata can decide how it should map those into errors (if it wishes), but yeah that bit we can make nicer :)